main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification3.scala

```diff
@@ -0,0 +1,308 @@
+      case Some(x) => SetFormula.mkElemSet(x)
+    }
+
+    case tpe: Type.AssocType => m.getForward(toAtom(tpe)) match {
+      case None => throw InternalCompilerException(s"Unexpected unbound associated type: '$t'.", t.loc)
+      case Some(x) => SetFormula.Cst(x)
+    }
+
+    case tpe@Type.Cst(TypeConstructor.Error(_, _), _) => m.getForward(toAtom(tpe)) match {
+      case None => throw InternalCompilerException(s"Unexpected unbound error type: '$t'.", t.loc)
+      case Some(x) => SetFormula.Var(x)
+    }
+
+    case Type.Apply(Type.Cst(TypeConstructor.Complement, _), tpe1, _) => SetFormula.mkCompl(toSetFormula(tpe1))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Union, _), tpe1, _), tpe2, _) => SetFormula.mkUnion(toSetFormula(tpe1), toSetFormula(tpe2))
+    case Type.Apply(Type.Apply(Type.Cst(TypeConstructor.Intersection, _), tpe1, _), tpe2, _) => SetFormula.mkInter(toSetFormula(tpe1), toSetFormula(tpe2))
+
+    case _ => throw InternalCompilerException(s"Unexpected type: '$t'.", t.loc)
+  }
+
+  /** Returns the Atom representation of the given Type. */
+  private def toAtom(t: Type)(implicit scope: Scope, renv: RigidityEnv): Atom = Type.eraseTopAliases(t) match {
+    case Type.Var(sym, _) => Atom.Var(sym)
+    case Type.Cst(TypeConstructor.Effect(sym), _) => Atom.Eff(sym)
+    case Type.AssocType(AssocTypeConstructor(sym, _), arg0, kind, _) =>
+      val arg = rigidToAtom(arg0)
+      Atom.Assoc(sym, arg, kind)
+    case Type.Cst(TypeConstructor.Error(id, kind), _) => Atom.Error(id, kind)
+    case tpe => throw InternalCompilerException(s"Unexpected non-atom type: $tpe", tpe.loc)
+  }
+
+  /** Returns the Atom representation of the given Type. */
+  private def rigidToAtom(t: Type)(implicit scope: Scope, renv: RigidityEnv): Atom = Type.eraseTopAliases(t) match {
+    case tpe@Type.Var(sym, _) =>
+      if (renv.isRigid(sym)) Atom.Var(sym)
+      else throw InternalCompilerException(s"Unexpected non-atom type: $tpe", tpe.loc)
+    // This is omitted to align with the "wrong" behaviour of the existing solver
+    //
+    //    case Type.Cst(TypeConstructor.Effect(sym), _) => Atom.Eff(sym)
+    //    case Type.AssocType(AssocTypeConstructor(sym, _), arg0, kind, _) =>
+    //      val arg = rigidToAtom(arg0)
+    //      Atom.Assoc(sym, arg, kind)
+    //    case Type.Cst(TypeConstructor.Error(id, kind), _) => Atom.Error(id, kind)
+    case tpe => throw InternalCompilerException(s"Unexpected non-atom type: $tpe", tpe.loc)
+  }
+
+  /** Returns a regular type substitution obtained from the given Boolean substitution `s`. */
+  private def fromSetSubst(s: SetSubstitution)(implicit m: Bimap[Atom, Int]): Substitution = {
+    Substitution(s.m.foldLeft(Map.empty[Symbol.KindedTypeVarSym, Type]) {
+      case (macc, (k, SetFormula.Var(x))) if k == x => macc
+      case (macc, (k, v)) =>
+        m.getBackward(k).get match {
+          // Case 1: A proper var. Add it to the substitution.
```

Comment:
What about rigidity?

Reply:
The rigidity checking was done at conversion, after that we know by the data which variables are flexible or not